timur/lib/client/jsx/components/view_editor/view_editor.jsx

```diff
@@ -0,0 +1,146 @@
+import React, {useState, useCallback, useEffect, shallowEqual} from 'react';
+import {useSelector, useDispatch, useStore} from 'react-redux';
+import {pushLocation, setLocation} from 'etna-js/actions/location_actions';
+import {
+  requestView,
+  requestAllViews,
+  deleteView,
+  saveNewView,
+  saveView
+} from '../../actions/view_actions';
+import DocumentWindow from '../document/document_window';
+import ViewScript from './views_script';
+
+// shallow selector
+export const useShallowEqualSelector = (selector) => {
+  return useSelector(selector, shallowEqual);
+};
+
+
+// Main component for viewing/editing views.
+
+const ViewEditor = ({view_id}) => {
+  let [editing, setEditing] = useState(useSelector(state => state.editing));
+  const dispatch = useDispatch();
+  let [view, setActualView] = useState(null);
```

Comment:
elsewhere we have `useActionInvoker` from `etna-js/hooks/useActionInvoker` - should we prefer useDispatch to this?

Reply:
They are similar.  If you checkout the code for `useActionInvoker` it is extremely similar to useDispatch except for adding on piece of functionality:

```
export function useActionInvoker() {
  const { store } = useContext(ReactReduxContext);
  return useCallback((action) => {
    if (typeof action === 'function') {
      return action(store.dispatch);
    }

    return store.dispatch(action);
  }, [store]);
}
```

Not only does it dispatch actions, it also can dispatch the results of functions that wrap taking the dispatcher (which we have scattered in many places), where as I think the default behavior of useDispatch assumes functions are themselves thunks (a reasonable assumption).  I think it'll take some time before we can consolidate our frontend patterns.

Reply:
I tried to use hooks over classes in React and instead of `connect` in Redux.